File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
using namespace std;
int tmp=0;
int dx[4]={0,0,1,-1};
int dy[4]={1,-1,0,0};
map<char, int> arrX;
map<char, int> arrY;
// e,w,s,n
vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer;
    vector<pair<int, int>> tmp;
    char container[4]={'E','W','S','N'};
    for(int i=0;i<4;i++){
        arrX[container[i]]=dx[i];
        arrY[container[i]]=dy[i];
    }
    for(int i=0;i<park.size();i++){
        for(int j=0;j<park[i].size();j++){
            if(park[i][j]=='S'){
                tmp.push_back({i,j});
                break;
            }
        }
    }
    
    int x=tmp.front().first;
    int y=tmp.front().second;
    for(int i=0;i<routes.size();i++){
        string direction=routes[i].substr(0, routes[i].find(" "));
        string destination=routes[i].substr(routes[i].find(" ")+1);
        char dir=direction[0];
        int des=stoi(destination);
        int nx=x+arrX[dir]*des;
        int ny=y+arrY[dir]*des;
        int nnx=x;
        int nny=y;
        bool flag=true;
        if(nx<0 || ny<0 || nx>=park.size() || ny>=park[0].size() || park[nx][ny]=='X') continue;
        for(int j=0;j<des;j++){
            nnx+=arrX[dir];
            nny+=arrY[dir];
            if(park[nnx][nny]=='X'){
                flag=false;
                break;
            }
        }
        if(flag==true){
            x=nx;
            y=ny;
            answer.clear();
            answer.push_back(nx);
            answer.push_back(ny);
        }
    }
    if(answer.size()==0){
        answer.push_back(x);
        answer.push_back(y);
    }
    return answer;
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
class Solution {
    public int[] solution(String[] park, String[] routes) {
        int[] answer = new int[2];
        int[] dx = {-1,1,0,0};
        int[] dy = {0,0,1,-1};
        // 북 남 동 서
        int sx = 0;
        int sy = 0;
        int rx = 0;
        int ry = 0;
        for(int i=0;i<park.length;i++){
            for(int j=0;j<park[i].length();j++){
                if(park[i].charAt(j)=='S'){
                    sx=i;
                    sy=j;
                    rx=i;
                    ry=j;
                    break;
                }
            }
        }
        for(String route: routes){
            boolean check = true;
            char dir = route.charAt(0);
            int dis = (int)(route.charAt(2))-48;
            int idx=0;
            if(dir=='N') idx=0;
            else if(dir=='S') idx=1;
            else if(dir=='W') idx=3;
            else if(dir=='E') idx=2;
            for(int i=0;i<dis;i++){
                int nx = sx+dx[idx];
                int ny = sy+dy[idx];
                if(nx>=0 && nx<park.length && ny>=0 && ny<park[0].length()){
                    System.out.println(park[nx].charAt(ny)+" "+nx+" "+ny);
                    if(park[nx].charAt(ny)=='X'){
                        check=false;
                        sx=rx;
                        sy=ry;
                        break;
                    }
                    else{
                        sx=nx;
                        sy=ny;
                    }
                }
                else{
                    check=false;
                    sx=rx;
                    sy=ry;
                    break;
                }
            }
            if(check==true){
                for(int i=0;i<dis;i++){
                    rx+=dx[idx];
                    ry+=dy[idx];
                }
            }
        }
        answer[0]=rx;
        answer[1]=ry;
        return answer;
    }
}